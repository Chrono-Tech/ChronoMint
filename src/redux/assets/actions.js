import contractsManagerDAO from 'dao/ContractsManagerDAO'

export const ASSETS_LIST = 'assets/LIST'

export const listAssets = () => async (dispatch) => {
  const dao = await contractsManagerDAO.getAssetsManagerDAO()
  dispatch({type: ASSETS_LIST, list: await dao.getList()})
}
