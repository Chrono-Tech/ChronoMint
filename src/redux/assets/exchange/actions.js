import type ExchangeManagerDAO from 'dao/ExchangeManagerDAO'
import type ExchangeModel from 'models/ExchangeModel'
import type TokenModel from 'models/TokenModel'
import contractsManagerDAO from 'dao/ContractsManagerDAO'

export const EXCHANGE_MANAGER_LIST = 'exchangeManager/LIST'
export const EXCHANGE_MANAGER_SET = 'exchangeManager/SET'
export const EXCHANGE_MANAGER_REMOVE = 'exchangeManager/REMOVE'

const getDAO = async (): Promise<ExchangeManagerDAO> => await contractsManagerDAO.getExchangeManagerDAO()

export const listExchanges = () => async (dispatch) => {
  const dao = await getDAO()
  dispatch({type: EXCHANGE_MANAGER_LIST, list: await dao.getList()})
}

export const addExchange = (address) => async () => {
  const exchange = new ExchangeModel({address, isFetching: true})
  dispatch({type: EXCHANGE_MANAGER_SET, exchange})
  try {
    const dao = await getDAO()
    await dao.addExchange(address)
  } catch (e) {
    // no rollback
  }
}

export const createExchange = (token: TokenModel) => async () => {
  try {
    const dao = await getDAO()
    await dao.createExchange(token)
  } catch (e) {
    // no rollback
  }
}

export const editExchange = (exchange: ExchangeModel, newExchange: ExchangeModel) => async () => {
  try {
    const dao = await getDAO()
    await dao.editExchange(exchange, newExchange)
  } catch (e) {
    // no rollback
  }
}

export const removeExchange = (exchange: ExchangeModel) => async () => {
  try {
    const dao = await getDAO()
    await dao.removeExchange(exchange)
  } catch (e) {
    // no rollback
  }
}

// TODO @bshevchenko: MINT-230 AssetsManager
// export const addOwner = (exchange: ExchangeModel, account) => async () => {
//   try {
//     const dao = await getDAO()
//     await dao.addOwner(exchange, account)
//   } catch (e) {
//     // no rollback
//   }
// }
//
// export const removeOwner = (exchange: ExchangeModel, account) => async () => {
//   try {
//     const dao = await getDAO()
//     await dao.removeOwner(exchange, account)
//   } catch (e) {
//     // no rollback
//   }
// }
