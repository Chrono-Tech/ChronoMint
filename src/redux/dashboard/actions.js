import LHTProxyDAO from '../../dao/LHTProxyDAO'
import UserStorageDAO from '../../dao/UserStorageDAO'

import {
  DASHBOARD_TOTAL_LHT_FETCH,
  DASHBOARD_TOTAL_LHT,
  DASHBOARD_TOTAL_MEMBERS_FETCH,
  DASHBOARD_TOTAL_MEMBERS
} from './reducer'

export const updateTotalLHT = () => (dispatch) => {
  dispatch({type: DASHBOARD_TOTAL_LHT_FETCH})
  return LHTProxyDAO.totalSupply()
    .then(balance => {
      dispatch({type: DASHBOARD_TOTAL_LHT, payload: balance})
    })
}

export const updateTotalMembers = () => (dispatch) => {
  dispatch({type: DASHBOARD_TOTAL_MEMBERS_FETCH})
  return UserStorageDAO.usersTotal()
    .then(number => {
      dispatch({type: DASHBOARD_TOTAL_MEMBERS, payload: number})
    })
}
