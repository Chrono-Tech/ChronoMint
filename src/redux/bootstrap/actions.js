import ls from 'utils/LocalStorage'
import { LOCAL_ID } from 'network/settings'

import {
  checkMetaMask, checkLocalSession, restoreLocalSession, createNetworkSession,
  checkTestRPC
} from '../network/actions'
import { login } from '../session/actions'

export const bootstrap = (relogin = true) => async (dispatch) => {
  dispatch(checkMetaMask())
  dispatch(checkTestRPC())

  if (!relogin) {
    return
  }

  const localAccount = ls.getLocalAccount()
  const isPassed = await dispatch(checkLocalSession(localAccount))
  if (isPassed) {
    await dispatch(restoreLocalSession(localAccount))
    dispatch(createNetworkSession(localAccount, LOCAL_ID, LOCAL_ID))
    dispatch(login(localAccount))
  } else {
    // eslint-disable-next-line
    console.warn('Can\'t restore local session')
  }
}
