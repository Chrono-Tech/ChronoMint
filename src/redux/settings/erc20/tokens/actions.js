import { I18n } from 'react-redux-i18n'
import { change } from 'redux-form'

import contractsManagerDAO from 'dao/ContractsManagerDAO'

import type AbstractFetchingModel from 'models/AbstractFetchingModel'
import type TokenModel from 'models/TokenModel'
import type TokenNoticeModel from 'models/notices/TokenNoticeModel'

import { notify } from 'redux/notifier/actions'
import { watchInitWallet, TIME } from 'redux/mainWallet/actions'

export const TOKENS_LIST = 'settings/TOKENS_LIST'
export const TOKENS_SET = 'settings/TOKENS_SET'
export const TOKENS_REMOVE = 'settings/TOKENS_REMOVE'
export const TOKENS_FORM = 'settings/TOKENS_FORM'
export const TOKENS_FORM_FETCH = 'settings/TOKENS_FORM_FETCH'

const setToken = (token: TokenModel) => ({ type: TOKENS_SET, token })
const removeToken = (token: TokenModel) => ({ type: TOKENS_REMOVE, token })

export const watchToken = (notice: TokenNoticeModel) => async (dispatch, getState) => {
  if (notice.isModified()) {
    for (const token: TokenModel of getState().get('settingsERC20Tokens').list.valueSeq().toArray()) {
      if (token.address() === notice.oldAddress()) {
        dispatch(removeToken(token))
        break
      }
    }
  }
  if (notice.isModified() || notice.isRemoved()) {
    if (getState().get('session').profile.tokens().toArray().includes(notice.token().address())
      || notice.token().symbol() === TIME) {
      dispatch(watchInitWallet())
    }
  }
  dispatch(notice.isRemoved() ?
    removeToken(notice.token()) : setToken(notice.token()))
  if (getState().get('session').isCBE) {
    dispatch(notify(notice))
  }
}

export const watchInitERC20Tokens = () => async dispatch => {
  const callback = notice => dispatch(watchToken(notice))

  const dao = await contractsManagerDAO.getERC20ManagerDAO()

  return Promise.all([
    dao.watchAdd(callback),
    dao.watchModify(callback),
    dao.watchRemove(callback),
  ])
}

export const listTokens = () => async dispatch => {
  const dao = await contractsManagerDAO.getERC20ManagerDAO()
  const list = await dao.getTokens()
  dispatch({ type: TOKENS_LIST, list })
}

export const formTokenLoadMetaData = async (token: TokenModel, dispatch, formName) => {
  dispatch({ type: TOKENS_FORM_FETCH })

  const managerDAO = await contractsManagerDAO.getERC20ManagerDAO()

  let dao
  try {
    dao = await contractsManagerDAO.getERC20DAO(token.address(), true)
  } catch (e) {
    dispatch({ type: TOKENS_FORM_FETCH, end: true })
    throw { address: I18n.t('settings.erc20.tokens.errors.invalidAddress') }
  }

  try {
    if (!token.decimals()) {
      dispatch(change(formName, 'decimals', dao.getDecimals()))
    }
    if (!token.symbol()) {
      dispatch(change(formName, 'symbol', dao.getSymbol()))
      token = token.setSymbol(dao.getSymbol())
    }
  } catch (e) {
    // eslint-disable-next-line
    console.warn('Load meta data error', e)
  }

  const symbolAddress = await managerDAO.getTokenAddressBySymbol(token.symbol())

  dispatch({ type: TOKENS_FORM_FETCH, end: true })

  if ((symbolAddress !== null && token.address() !== symbolAddress) || token.symbol().toUpperCase() === 'ETH') {
    throw { symbol: I18n.t('settings.erc20.tokens.errors.symbolInUse') }
  }
}

export const addToken = (token: TokenModel | AbstractFetchingModel) => async dispatch => {
  dispatch(setToken(token.isFetching(true)))
  const dao = await contractsManagerDAO.getERC20ManagerDAO()
  try {
    await dao.addToken(token)
  } catch (e) {
    dispatch(removeToken(token))
  }
}

export const modifyToken = (oldToken: TokenModel | AbstractFetchingModel, newToken: TokenModel) => async dispatch => {
  dispatch(setToken(oldToken.isFetching(true)))
  const dao = await contractsManagerDAO.getERC20ManagerDAO()
  try {
    await dao.modifyToken(oldToken, newToken)
    dispatch(removeToken(oldToken))
  } catch (e) {
    dispatch(setToken(oldToken.isFetching(false)))
  }
}

export const revokeToken = (token: TokenModel | AbstractFetchingModel) => async dispatch => {
  dispatch(setToken(token.isFetching(true)))
  const dao = await contractsManagerDAO.getERC20ManagerDAO()
  try {
    await dao.removeToken(token)
  } catch (e) {
    dispatch(setToken(token.isFetching(false)))
  }
}
