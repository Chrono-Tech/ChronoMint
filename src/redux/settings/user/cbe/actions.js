import { change } from 'redux-form/immutable'
import type AbstractFetchingModel from 'models/AbstractFetchingModel'
import type CBEModel from 'models/CBEModel'
import type CBENoticeModel from 'models/notices/CBENoticeModel'
import contractsManagerDAO from 'dao/ContractsManagerDAO'
import { notify } from 'redux/notifier/actions'
import { FORM_CBE_ADDRESS } from 'components/dialogs/CBEAddressDialog'

export const CBE_LIST = 'settings/CBE_LIST'
export const CBE_FORM = 'settings/CBE_FORM'
export const CBE_SET = 'settings/CBE_SET'
export const CBE_LOADING = 'settings/CBE_LOADING'
export const CBE_REMOVE = 'settings/CBE_REMOVE'

export const setCBE = (cbe: CBEModel) => ({type: CBE_SET, cbe})
export const removeCBE = (cbe: CBEModel) => ({type: CBE_REMOVE, cbe})

export const listCBE = () => async (dispatch) => {
  const dao = await contractsManagerDAO.getUserManagerDAO()
  const list = await dao.getCBEList()
  dispatch({type: CBE_LIST, list})
}

export const formCBELoadName = (account) => async (dispatch) => {
  dispatch({type: CBE_LOADING, isLoading: true})
  const dao = await contractsManagerDAO.getUserManagerDAO()
  const profile = await dao.getMemberProfile(account)
  dispatch({type: CBE_LOADING, isLoading: false})
  dispatch(change(FORM_CBE_ADDRESS, 'name', profile.name()))
}

export const addCBE = (cbe: CBEModel | AbstractFetchingModel) => async (dispatch) => {
  dispatch(setCBE(cbe.isFetching(true)))
  const dao = await contractsManagerDAO.getUserManagerDAO()
  try {
    await dao.addCBE(cbe)
  } catch (e) {
    dispatch(removeCBE(cbe))
  }
}

export const revokeCBE = (cbe: CBEModel | AbstractFetchingModel) => async (dispatch) => {
  dispatch(setCBE(cbe.isFetching(true)))
  const dao = await contractsManagerDAO.getUserManagerDAO()
  try {
    await dao.revokeCBE(cbe)
  } catch (e) {
    dispatch(setCBE(cbe))
  }
}

export const watchCBE = (notice: CBENoticeModel) => dispatch => {
  dispatch(notify(notice))
  dispatch(notice.isRevoked() ? removeCBE(notice.cbe()) : setCBE(notice.cbe()))
}

export const watchInitCBE = () => async (dispatch) => {
  const dao = await contractsManagerDAO.getUserManagerDAO()
  return dao.watchCBE((notice) => dispatch(watchCBE(notice)))
}
