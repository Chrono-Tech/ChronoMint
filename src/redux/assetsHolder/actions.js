import { DUCK_SESSION } from '@/redux/session/actions'
import contractsManagerDAO from 'dao/ContractsManagerDAO'
import TIMEHolderDAO from 'dao/TIMEHolderDAO'
import Amount from 'models/Amount'
import ApprovalNoticeModel from 'models/notices/ApprovalNoticeModel'
import TokenModel from 'models/tokens/TokenModel'
import { DUCK_TOKENS } from 'redux/tokens/actions'
import tokenService, { EVENT_NEW_TOKEN } from 'services/TokenService'

export const DUCK_ASSETS_HOLDER = 'assetsHolder'

export const TIME_HOLDER_INIT = 'timeHolder/INIT'
export const TIME_HOLDER_TIME_ADDRESS = 'timeHolder/timeAddress'
export const TIME_HOLDER_WALLET_ADDRESS = 'timeHolder/timeHolderWalletAddress'
export const TIME_HOLDER_ADDRESS = 'timeHolder/timeHolderAddress'

let timeHolderDAO: TIMEHolderDAO = null

const subscribeOnTokens = (token: TokenModel) => async (dispatch, getState) => {
  if (token.symbol() !== 'TIME') {
    return
  }

  dispatch({ type: TIME_HOLDER_TIME_ADDRESS, address: token.address() })

  // init
  const [ timeHolderAddress, timeHolderWalletAddress ] = await Promise.all([
    timeHolderDAO.getAddress(),
    timeHolderDAO.getWalletAddress(),
  ])

  dispatch({ type: TIME_HOLDER_ADDRESS, address: timeHolderAddress })
  dispatch({ type: TIME_HOLDER_WALLET_ADDRESS, address: timeHolderWalletAddress })

  // TODO @dkchv: review again
  const contractNames = {}
  contractNames[ timeHolderAddress ] = 'TIME Holder'
  ApprovalNoticeModel.setContractNames(contractNames)

  // update allowance
  const timeDAO = tokenService.getDAO(token.id())
  const { account } = getState().get(DUCK_SESSION)
  const [ timeHolderAllowance, timeHolderWalletAllowance ] = await Promise.all([
    timeDAO.getAccountAllowance(timeHolderAddress, account),
    timeDAO.getAccountAllowance(timeHolderWalletAddress, account),
  ])

  console.log('--assetHolder: ', +timeHolderAllowance, +timeHolderWalletAllowance)

  // token = token
  //   .setAllowance(timeHolderAddress, timeHolderAllowance)
  //   .setAllowance(timeHolderWalletAddress, timeHolderWalletAllowance)
}

export const initAssetsHolder = () => async (dispatch, getState) => {
  if (getState().get(DUCK_ASSETS_HOLDER).isInited()) {
    return
  }
  dispatch({ type: TIME_HOLDER_INIT, inInited: true })

  const { account } = getState().get(DUCK_SESSION)
  timeHolderDAO = await contractsManagerDAO.getTIMEHolderDAO()
  const [ assets, deposit ] = await Promise.all([
    timeHolderDAO.getSharesContract(),
    timeHolderDAO.getDeposit(account),
  ])
  console.log('--deposit', assets, +deposit)

  // subscribe to tokens
  const callback = (token) => dispatch(subscribeOnTokens(token))
  tokenService.on(EVENT_NEW_TOKEN, callback)
  // fetch for existing tokens
  const tokens = getState().get(DUCK_TOKENS)
  tokens.list().forEach(callback)
}

export const depositAsset = (amount: Amount, token: TokenModel) => async (dispatch) => {
  // const amountBN = new BigNumber(amount)

  // dispatch(balanceMinus(amountBN, token))
  const assetDAO = tokenService.getDAO(token)

  try {
    await timeHolderDAO.deposit(amount, assetDAO)
  } catch (e) {
    // eslint-disable-next-line
    console.error('deposit error', e.message)
  } finally {
    // compensation for update in watchTransfer
    // dispatch(balancePlus(amount, token))
  }
}

export const withdrawAsset = (amount: Amount, token: TokenModel) => async () => {
  // const amountBN = new BigNumber(amount)

  // dispatch(depositMinus(amountBN))
  const assetDAO = tokenService.getDAO(token)

  try {
    await timeHolderDAO.withdraw(amount, assetDAO)
  } catch (e) {
    // eslint-disable-next-line
    console.error('withdraw error', e.message)
  } finally {
    // dispatch(depositPlus(amountBN))
  }
}
