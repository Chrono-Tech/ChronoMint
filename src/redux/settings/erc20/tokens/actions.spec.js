import { store } from 'specsInit'
import contractsManagerDAO from 'dao/ContractsManagerDAO'
import fakeCoinDAO from 'dao/FakeCoinDAO'
import type AbstractFetchingModel from 'models/AbstractFetchingModel'
import TokenModel from 'models/TokenModel'
import type TokenNoticeModel from 'models/notices/TokenNoticeModel'
import { TIME } from 'redux/mainWallet/actions'
import * as a from './actions'

let token: TokenModel | AbstractFetchingModel
let newToken: TokenModel | AbstractFetchingModel

describe('settings erc20 actions', () => {
  afterEach(async (done) => {
    await setTimeout(() => {
      done()
    }, 3000)
  })
  it('should dispatch changes for decimals and symbol fields for valid ERC20', async () => {
    await a.formTokenLoadMetaData(new TokenModel({
      address: await fakeCoinDAO.getAddress(),
    }), store.dispatch, 'TEST')

    expect(store.getActions()[ 1 ].payload).toMatchSnapshot()
    expect(store.getActions()[ 2 ].payload).toMatchSnapshot()
  })

  it('should throw symbolInUse error if token symbol is already in use', async () => {
    try {
      await a.formTokenLoadMetaData(new TokenModel({
        address: await fakeCoinDAO.getAddress(),
        symbol: TIME,
      }), store.dispatch, 'TEST')
    } catch (e) {
      expect(e).toMatchSnapshot()
    }
  })

  it('should add token', async (done) => {
    token = new TokenModel({
      symbol: 'FAKE',
      name: 'Fake',
      address: await fakeCoinDAO.getAddress(),
      decimals: 3,
      url: 'http://fakecoin.io',
      icon: 'QmaJUTXy22JyWnjoy3d34j89ifzqdwKZC2QziKWMja7GsS',
    })

    const dao = await contractsManagerDAO.getERC20ManagerDAO()
    await dao.watchAdd((notice: TokenNoticeModel) => {
      expect(store.getActions()[ 0 ].token.address(null)).toMatchSnapshot()
      expect(notice.isRemoved()).toMatchSnapshot()
      expect(notice.token().address(null)).toMatchSnapshot()
      done()
    })

    await store.dispatch(a.addToken(token))
  })

  it('should list tokens', async () => {
    await store.dispatch(a.listTokens())
    expect(store.getActions()[ 0 ].list.get(token.symbol()).address(null)).toMatchSnapshot()
  })

  it('should throw invalidAddress error if token is not valid ERC20', async () => {
    try {
      await a.formTokenLoadMetaData(new TokenModel({
        address: '0xc38f003c0a14a05f11421d793edc9696a25cb2b3',
      }), store.dispatch, 'TEST')
    } catch (e) {
      expect(e).toMatchSnapshot()
    }
  })

  it('should modify token', async (done) => {
    newToken = token.setSymbol('FAKE')
    const result = await store.dispatch(a.modifyToken(token, newToken))
    expect(result).toMatchSnapshot()
    done()
  })

  it('should remove token', async (done) => {
    const dao = await contractsManagerDAO.getERC20ManagerDAO()
    await dao.watchRemove((notice: TokenNoticeModel) => {
      expect(store.getActions()).toMatchSnapshot()
      expect(notice.isRemoved()).toMatchSnapshot()
      expect(notice.token()).toMatchSnapshot()
      done()
    })
    await store.dispatch(a.revokeToken(newToken))
  })
})
