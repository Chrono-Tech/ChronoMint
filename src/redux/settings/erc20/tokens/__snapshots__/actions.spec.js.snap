// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`settings erc20 actions should add token 1`] = `
Object {
  "address": "0xb2bc23dbafe9557819c01cca5c7f699a63e3d0a3",
  "allowance": Object {},
  "balance": "0",
  "dao": null,
  "decimals": 3,
  "fee": null,
  "feeAddress": null,
  "icon": "QmaJUTXy22JyWnjoy3d34j89ifzqdwKZC2QziKWMja7GsS",
  "id": null,
  "isFailed": false,
  "isFetched": false,
  "isFetching": true,
  "isOptional": true,
  "isPending": false,
  "isReissuable": null,
  "managersList": null,
  "name": "Fake",
  "platform": null,
  "symbol": "FAKE",
  "timestamp": null,
  "totalSupply": "0",
  "transactionHash": null,
  "url": "http://fakecoin.io",
  "withFee": null,
}
`;

exports[`settings erc20 actions should add token 2`] = `false`;

exports[`settings erc20 actions should add token 3`] = `
Object {
  "address": "0xb2bc23dbafe9557819c01cca5c7f699a63e3d0a3",
  "allowance": Object {},
  "balance": "0",
  "dao": null,
  "decimals": 3,
  "fee": null,
  "feeAddress": null,
  "icon": "QmaJUTXy22JyWnjoy3d34j89ifzqdwKZC2QziKWMja7GsS",
  "id": null,
  "isFailed": false,
  "isFetched": false,
  "isFetching": false,
  "isOptional": true,
  "isPending": false,
  "isReissuable": null,
  "managersList": null,
  "name": "Fake",
  "platform": null,
  "symbol": "FAKE",
  "timestamp": null,
  "totalSupply": "0",
  "transactionHash": null,
  "url": "http://fakecoin.io",
  "withFee": null,
}
`;

exports[`settings erc20 actions should dispatch changes for decimals and symbol fields for valid ERC20 1`] = `4`;

exports[`settings erc20 actions should dispatch changes for decimals and symbol fields for valid ERC20 2`] = `"FAKE"`;

exports[`settings erc20 actions should list tokens 1`] = `
Object {
  "address": "0xb2bc23dbafe9557819c01cca5c7f699a63e3d0a3",
  "allowance": Object {},
  "balance": "0",
  "dao": null,
  "decimals": 3,
  "fee": null,
  "feeAddress": null,
  "icon": "QmaJUTXy22JyWnjoy3d34j89ifzqdwKZC2QziKWMja7GsS",
  "id": null,
  "isFailed": false,
  "isFetched": true,
  "isFetching": false,
  "isOptional": true,
  "isPending": false,
  "isReissuable": null,
  "managersList": null,
  "name": "Fake",
  "platform": null,
  "symbol": "FAKE",
  "timestamp": null,
  "totalSupply": "0",
  "transactionHash": null,
  "url": "http://fakecoin.io",
  "withFee": null,
}
`;

exports[`settings erc20 actions should modify token 1`] = `
Array [
  Object {
    "token": Object {
      "address": "0xb2bc23dbafe9557819c01cca5c7f699a63e3d0a3",
      "allowance": Object {},
      "balance": "0",
      "dao": null,
      "decimals": 3,
      "fee": null,
      "feeAddress": null,
      "icon": "QmaJUTXy22JyWnjoy3d34j89ifzqdwKZC2QziKWMja7GsS",
      "id": null,
      "isFailed": false,
      "isFetched": false,
      "isFetching": true,
      "isOptional": true,
      "isPending": false,
      "isReissuable": null,
      "managersList": null,
      "name": "Fake",
      "platform": null,
      "symbol": "FAKE",
      "timestamp": null,
      "totalSupply": "0",
      "transactionHash": null,
      "url": "http://fakecoin.io",
      "withFee": null,
    },
    "type": "settings/TOKENS_SET",
  },
  Object {
    "token": Object {
      "address": "0xb2bc23dbafe9557819c01cca5c7f699a63e3d0a3",
      "allowance": Object {},
      "balance": "0",
      "dao": null,
      "decimals": 3,
      "fee": null,
      "feeAddress": null,
      "icon": "QmaJUTXy22JyWnjoy3d34j89ifzqdwKZC2QziKWMja7GsS",
      "id": null,
      "isFailed": false,
      "isFetched": false,
      "isFetching": false,
      "isOptional": true,
      "isPending": false,
      "isReissuable": null,
      "managersList": null,
      "name": "Fake",
      "platform": null,
      "symbol": "FAKE",
      "timestamp": null,
      "totalSupply": "0",
      "transactionHash": null,
      "url": "http://fakecoin.io",
      "withFee": null,
    },
    "type": "settings/TOKENS_REMOVE",
  },
]
`;

exports[`settings erc20 actions should modify token 2`] = `true`;

exports[`settings erc20 actions should modify token 3`] = `
Object {
  "address": "0xb2bc23dbafe9557819c01cca5c7f699a63e3d0a3",
  "allowance": Object {},
  "balance": "0",
  "dao": null,
  "decimals": 3,
  "fee": null,
  "feeAddress": null,
  "icon": "QmaJUTXy22JyWnjoy3d34j89ifzqdwKZC2QziKWMja7GsS",
  "id": null,
  "isFailed": false,
  "isFetched": false,
  "isFetching": false,
  "isOptional": true,
  "isPending": false,
  "isReissuable": null,
  "managersList": null,
  "name": "Fake",
  "platform": null,
  "symbol": "CAKE",
  "timestamp": null,
  "totalSupply": "0",
  "transactionHash": null,
  "url": "http://fakecoin.io",
  "withFee": null,
}
`;

exports[`settings erc20 actions should modify token 4`] = `"0xb2bc23dbafe9557819c01cca5c7f699a63e3d0a3"`;

exports[`settings erc20 actions should remove token 1`] = `
Array [
  Object {
    "token": Object {
      "address": "0xb2bc23dbafe9557819c01cca5c7f699a63e3d0a3",
      "allowance": Object {},
      "balance": "0",
      "dao": null,
      "decimals": 3,
      "fee": null,
      "feeAddress": null,
      "icon": "QmaJUTXy22JyWnjoy3d34j89ifzqdwKZC2QziKWMja7GsS",
      "id": null,
      "isFailed": false,
      "isFetched": false,
      "isFetching": true,
      "isOptional": true,
      "isPending": false,
      "isReissuable": null,
      "managersList": null,
      "name": "Fake",
      "platform": null,
      "symbol": "CAKE",
      "timestamp": null,
      "totalSupply": "0",
      "transactionHash": null,
      "url": "http://fakecoin.io",
      "withFee": null,
    },
    "type": "settings/TOKENS_SET",
  },
]
`;

exports[`settings erc20 actions should remove token 2`] = `true`;

exports[`settings erc20 actions should remove token 3`] = `
Object {
  "address": "0xb2bc23dbafe9557819c01cca5c7f699a63e3d0a3",
  "allowance": Object {},
  "balance": "0",
  "dao": null,
  "decimals": 3,
  "fee": null,
  "feeAddress": null,
  "icon": "QmaJUTXy22JyWnjoy3d34j89ifzqdwKZC2QziKWMja7GsS",
  "id": null,
  "isFailed": false,
  "isFetched": false,
  "isFetching": false,
  "isOptional": true,
  "isPending": false,
  "isReissuable": null,
  "managersList": null,
  "name": "Fake",
  "platform": null,
  "symbol": "CAKE",
  "timestamp": null,
  "totalSupply": "0",
  "transactionHash": null,
  "url": "http://fakecoin.io",
  "withFee": null,
}
`;

exports[`settings erc20 actions should throw invalidAddress error if token is not valid ERC20 1`] = `
Object {
  "address": "Invalid Address",
}
`;

exports[`settings erc20 actions should throw symbolInUse error if token symbol is already in use 1`] = `
Object {
  "symbol": "Symbol In Use",
}
`;
